modules/platforms/python/pygridgain/client.py

```diff
@@ -185,36 +149,51 @@ def connect(self, *args):
         else:
             raise ConnectionError('Connection parameters are not valid.')
 
-        nodes = iter(nodes)
+        # the following code is quite twisted, because the protocol version
+        # is initially unknown
+
         # TODO: open first node in foregroung, others − in background
+        for i, node in enumerate(nodes):
+            host, port = node
+            conn = Connection(self, **self._connection_args)
+            conn.host = host
+            conn.port = port
 
-        first_node = Connection(self, **self._connection_args)
+            try:
+                if (
+                    self.protocol_version is None
+                    or self.protocol_version >= (1, 4, 0)
+                ):
+                    # open connection before adding to the pool
+                    conn.connect(host, port)
+
+                    # now we have the protocol version
+                    if self.protocol_version < (1, 4, 0):
+                        # do not try to open more nodes
+                        self._current_node = i
+                    else:
+                        # take a chance to schedule the reconnection
+                        # for all the failed connections, that was probed
+                        # before this
+                        for failed_node in self._nodes[:i]:
```

Comment:
What about not-failed nodes, which was not attempted to connect yet?

Reply:
They will stay closed until then. I suppose the scheduling part will not take long, since the reconnection itself it will be performed in background threads.

The point of the code is that we should have at least one opened connection to determine the version of thin protocol. Background reconnection only have sense for 1.4.

Reply:
Why does it only have sense for 1.4?

Reply:
With thin protocol version 1.3 and below, client use only one node. To restore its function after the connection breakdown, we need to connect to just one node. User must wait for just that to happen. It is a simple algorithm, and that is how the client was implemented from the start.

It is the Best Effort Affinity feature that makes client benefit from multiple open connections. But the client also can work with some of the connections failing. Then it makes sense to reopen the failed connections in the background, so the user does not have to wait for it, but just use a suboptimal path for some requests.

Now when you ask why, I am starting to think that it wouldn't hurt to always have a spare connection, so the user of protocol 1.3- also does not have to wait for the reconnection. But the benefit from fast reopening of just one usable connection will definitely be not that great as with reopening of virtually unlimited number of connections, as of v1.4.